pkg/apis/serving/v1alpha1/revision_lifecycle.go

```diff
@@ -270,3 +271,84 @@ func (r *Revision) GetLastPinned() (time.Time, error) {
 func (rs *RevisionStatus) duck() *duckv1beta1.Status {
 	return &rs.Status
 }
+
+// PropagateDeploymentStatus takes the Deployment status and applies its values
+// to the Revision status.
+func (rs *RevisionStatus) PropagateDeploymentStatus(original *appsv1.DeploymentStatus) {
+	ds := transformDeploymentStatus(original)
+	cond := ds.getCondition(deploymentConditionReady)
+	if cond == nil {
+		return
+	}
+
+	switch cond.Status {
+	case corev1.ConditionUnknown:
+		revCondSet.Manage(rs).MarkUnknown(RevisionConditionResourcesAvailable, cond.Reason, cond.Message)
+	case corev1.ConditionTrue:
+		revCondSet.Manage(rs).MarkTrue(RevisionConditionResourcesAvailable)
+	case corev1.ConditionFalse:
+		revCondSet.Manage(rs).MarkFalse(RevisionConditionResourcesAvailable, cond.Reason, cond.Message)
+	}
+}
+
+var depCondSet = apis.NewLivingConditionSet(
+	deploymentConditionProgressing,
+	deploymentConditionReplicaSetReady,
+)
+
+const (
+	// deploymentConditionReady means the underlying deployment is ready.
+	deploymentConditionReady = apis.ConditionReady
+	// deploymentConditionReplicaSetready inverts the underlying deployment's
+	// ReplicaSetFailure condition.
+	deploymentConditionReplicaSetReady apis.ConditionType = "ReplicaSetReady"
+	// deploymentConditionProgressing reflects the underlying deployment's
+	// Progressing condition.
+	deploymentConditionProgressing apis.ConditionType = "Progressing"
+)
+
+type deploymentStatus struct {
+	duckv1beta1.Status `json:",inline"`
+}
+
+func (ds *deploymentStatus) getCondition(t apis.ConditionType) *apis.Condition {
+	return depCondSet.Manage(ds).GetCondition(t)
+}
+
+func (ds *deploymentStatus) initializeConditions() {
+	depCondSet.Manage(ds).InitializeConditions()
+	// The absence of this condition means no failure has occured.
+	depCondSet.Manage(ds).MarkTrue(deploymentConditionReplicaSetReady)
+}
+
+// transformDeploymentStatus transforms the kubernetes DeploymentStatus into a
+// duckv1beta1.Status that uses ConditionSets to propagate failures and expose
+// a top-level happy state, per our condition conventions.
+func transformDeploymentStatus(ds *appsv1.DeploymentStatus) *deploymentStatus {
+	s := &deploymentStatus{}
+	s.initializeConditions()
+	for _, cond := range ds.Conditions {
+		// TODO(jonjohnsonjr): Should we care about appsv1.DeploymentAvailable here?
+		switch cond.Type {
+		case appsv1.DeploymentProgressing:
+			switch cond.Status {
+			case corev1.ConditionUnknown:
+				depCondSet.Manage(s).MarkUnknown(deploymentConditionProgressing, cond.Reason, cond.Message)
+			case corev1.ConditionTrue:
+				depCondSet.Manage(s).MarkTrue(deploymentConditionProgressing)
+			case corev1.ConditionFalse:
+				depCondSet.Manage(s).MarkFalse(deploymentConditionProgressing, cond.Reason, cond.Message)
+			}
+		case appsv1.DeploymentReplicaFailure:
```

Comment:
Does this condition only have general failures across the board or is it triggered by even a single replica failure? I'm mostly making sure we don't mark the Revision as unready and subsequently cause backlash in the routing layer if only a single replica fails.

Reply:
As far as I can tell, this just surfaces any error that happens when a ReplicaSet attempts to create or delete a pod, see [here](https://github.com/kubernetes/kubernetes/blob/a3ccea9d8743f2ff82e41b6c2af6dc2c41dc7b10/pkg/controller/replicaset/replica_set_utils.go#L108-L120).

From the Deployment's [description](https://godoc.org/k8s.io/api/apps/v1#DeploymentReplicaFailure):
> // ReplicaFailure is added in a deployment when one of its pods fails to be created
> // or deleted.

And the ReplicaSet [description](https://godoc.org/k8s.io/api/apps/v1#ReplicaSetReplicaFailure) expands on that:
> // ReplicaSetReplicaFailure is added in a replica set when one of its pods fails to be 
> // due to insufficient quota, limit ranges, pod security policy, node selectors, etc. or 
> // due to kubelet being down or finalizers are failing.

E.g. a pod crashing wouldn't trigger this, but being unable to recreate a new pod would.

Re: the routing layer, my understanding is that a Revision being Ready=False wouldn't blackhole a revision that's already routed, but would prevent a revision from being routed in the first place.